libp2p/connmanager.nim

```diff
@@ -331,35 +345,120 @@ proc storeConn*(c: ConnManager, conn: Connection) =
   ##
 
   if isNil(conn):
-    raise newException(CatchableError, "connection cannot be nil")
+    raise newException(CatchableError, "Connection cannot be nil")
 
-  if conn.closed() or conn.atEof():
-    trace "Can't store dead connection", conn
-    raise newException(CatchableError, "can't store dead connection")
+  if conn.closed or conn.atEof:
+    raise newException(CatchableError, "Connection closed or EOF")
 
   if isNil(conn.peerInfo):
-    raise newException(CatchableError, "empty peer info")
+    raise newException(CatchableError, "Empty peer info")
+
+  # Launch on close listener
+  # All the errors are handled inside `onClose()` procedure.
+  asyncSpawn c.onClose(conn)
 
   let peerId = conn.peerInfo.peerId
-  if c.conns.getOrDefault(peerId).len > c.maxConns:
-    debug "too many connections",
+  if c.conns.getOrDefault(peerId).len > c.maxPeerConns:
+    debug "Too many connections for peer",
       conn, conns = c.conns.getOrDefault(peerId).len
 
-    raise newTooManyConnections()
+    raise newTooManyConnectionsError()
 
   if peerId notin c.conns:
     c.conns[peerId] = initHashSet[Connection]()
 
   c.conns[peerId].incl(conn)
-
-  # Launch on close listener
-  # All the errors are handled inside `onClose()` procedure.
-  asyncSpawn c.onClose(conn)
   libp2p_peers.set(c.conns.len.int64)
 
   trace "Stored connection",
     conn, direction = $conn.dir, connections = c.conns.len
 
+proc trackConn(c: ConnManager,
+               provider: ConnProvider,
+               dir: Direction):
+               Future[Connection] {.async.} =
+  var conn: Connection
+  try:
+    conn = await provider()
+
+    if isNil(conn):
+      return
+
+    conn.dir = dir
+    trace "Got connection", conn, dir = $dir
+
+    proc semaphoreMonitor() {.async.} =
+      try:
+        await conn.join()
+      except CatchableError as exc:
+        trace "Exception in semaphore monitor, ignoring", exc = exc.msg
+
+      if conn.dir == Direction.In:
+        c.inConnSemaphore.release()
+      else:
+        c.outConnSemaphore.release()
+
+    asyncSpawn semaphoreMonitor()
+  except CatchableError as exc:
+    trace "Exception tracking connection", exc = exc.msg
+    if not isNil(conn):
+      await conn.close()
+
+    raise exc
+
+  return conn
+
+proc trackIncomingConn*(c: ConnManager,
+                        provider: ConnProvider):
+                        Future[Connection] {.async.} =
+  ## await for a connection slot before attempting
+  ## to call the connection provider
+  ##
+
+  var conn: Connection
+  try:
+    trace "Tracking incoming connection"
+    await c.inConnSemaphore.acquire()
+    conn = await c.trackConn(provider, Direction.In)
+    if isNil(conn):
+      trace "Couldn't acquire connection, releasing semaphore slot", dir = $Direction.In
+      c.inConnSemaphore.release()
+  except CatchableError as exc:
+    trace "Exception tracking connection", exc = exc.msg
+    c.inConnSemaphore.release()
```

Comment:
the release should not happen if an exception was raised before the semaphore was acquired - for example if `acquire` was cancelled before being acquired

Reply:
Yes, this needs to be fixed once the semaphore cancellations are merged, without it, unless explicitly releasing on cancelation this branch will break.